src/client/dfuse/pil4dfs/hook.c

```diff
@@ -0,0 +1,981 @@
+
+	if (found_libc == 0 || is_uninstalled == 1)
+		return;
+
+	for (iBlk = 0; iBlk < num_patch_blk; iBlk++) {
+		tramp_list = (struct trampoline_t *)(patch_blk_list[iBlk].patch_addr);
+
+		for (iFunc = 0; iFunc < patch_blk_list[iBlk].num_trampoline; iFunc++) {
+			/* fast mod */
+			pbaseOrg = (void *)((long int)(tramp_list[iFunc].addr_org_func) & mask);
+			MemSize_Modify = determine_mem_block_size(
+			    (const void *)(tramp_list[iFunc].addr_org_func));
+
+			if (pbaseOrg == NULL)
+				continue;
+			if (mprotect(pbaseOrg, MemSize_Modify,
+				     PROT_READ | PROT_WRITE | PROT_EXEC) != 0) {
+				printf("Error in executing p_mp().\n");
+				exit(1);
+			}
+			/* save original code for uninstall */
+			memcpy(tramp_list[iFunc].addr_org_func, tramp_list[iFunc].org_code, 5);
+			if (mprotect(pbaseOrg, MemSize_Modify, PROT_READ | PROT_EXEC) != 0) {
+				printf("Error in executing p_mp().\n");
+				exit(1);
+			}
+		}
+	}
+
+	for (i = 0; i < num_patch_blk; i++) {
+		if (patch_blk_list[i].patch_addr) {
+			if (munmap(patch_blk_list[i].patch_addr, MIN_MEM_SIZE))
+				perror("munmap");
+			patch_blk_list[i].patch_addr = 0;
+		}
+	}
+
+	is_uninstalled = 1;
+}
+
+/*
+ * query_lib_name_in_list - Query the index of the name of a library in lib_name_list[].
+ *   @lib_name_str: The library name
+ * Returns:
+ *   The index in lib_name_list[]. (-1) means not found in the list.
+ */
+static int
+query_lib_name_in_list(const char *lib_name_str)
+{
+	int i;
+
+	/* Try exact match first */
+	for (i = 0; i < num_lib_in_map; i++) {
+		if (strcmp(lib_name_str, lib_name_list[i]) == 0)
+			return i;
+	}
+	/* Try partial match */
+	for (i = 0; i < num_lib_in_map; i++) {
+		if (strstr(lib_name_list[i], lib_name_str))
+			return i;
+	}
+	return (-1);
+}
+
+/*
+ * query_registered_module - Query the index of the name of a library in module_list[].
+ *   @lib_name_str: The library name
+ * Returns:
+ *   The index in module_list[]. (-1) means not found in the list.
+ */
+static int
+query_registered_module(const char *lib_name_str)
+{
+	int i;
+
+	for (i = 0; i < num_module; i++) {
+		if (strcmp(lib_name_str, module_list[i].module_name) == 0)
+			return i;
+	}
+	return (-1);
+}
+
+/*
+ * get_position_of_next_line - Determine the offset of the next new line in a string buffer.
+ *   @buff: The string buffer
+ *   @pos_start: The starting offset to search
+ *   @max_buff_size: The max length of buff[]
+ * Returns:
+ *   The offset of the next new line. (-1) means reaching the end of buffer.
+ */
+static int
+get_position_of_next_line(const char buff[], const int pos_start, const int max_buff_size)
+{
+	int i = pos_start;
+
+	while (i < max_buff_size) {
+		/* A new line */
+		if (buff[i] == 0xA) {
+			i++;
+			return ((i >= max_buff_size) ? (-1) : (i));
+		}
+		i++;
+	}
+	return (-1);
+}
+
+/**
+ * The max size of read "/proc/self/maps". This size set here should be sufficient
+ * for normal applications.
+ */
+#define MAX_MAP_SIZE (524288)
```

Comment:
Any justification for this claim?  What happens if it's not sufficent?

Reply:
I will change the code to always read the whole file.

Reply:
Fixed with adaptive reading the whole file. Thank you!